[KubeRay][Doc] KubeRay doc migration (2/n) (#39074)
ray-project/ray: doc/source/cluster/kubernetes/user-guides/rayservice.md

curl -X POST -H 'Content-Type: application/json' rayservice-sample-serve-svc:8000/fruit/ -d '["MANGO", 2]'
# [Expected output]: 8
```

## Step 8: Zero downtime upgrade for Ray clusters

In Step 7, modifying `serveConfigV2` will not trigger a zero downtime upgrade for Ray clusters.
Instead, it will reapply the new configurations to the existing RayCluster.
However, if you modify `spec.rayClusterConfig` in the RayService YAML file, it will trigger a zero downtime upgrade for Ray clusters.
To elaborate, RayService will temporarily create a new RayCluster and wait for it to be ready, then switch traffic to the new RayCluster by updating the selector of the head service managed by RayService (i.e. `rayservice-sample-head-svc`) and terminate the old one.

To elaborate, during the zero downtime upgrade process, RayService will create a new RayCluster temporarily and wait for it to become ready. 
Once the new RayCluster is ready, RayService will update the selector of the head service managed by RayService (i.e., `rayservice-sample-head-svc`) to point to the new RayCluster to switch the traffic to the new RayCluster. 
Finally, the old RayCluster will be terminated.

> There are certain exceptions that will not trigger a zero downtime upgrade.

Reviewer: ```suggestion
Certain exceptions don't trigger a zero downtime upgrade.
```

> Currently, only the fields managed by Ray Autoscaler, `replicas` and `scaleStrategy.workersToDelete`, will not trigger a zero downtime upgrade.

Reviewer: ```suggestion
Only the fields managed by Ray Autoscaler, `replicas` and `scaleStrategy.workersToDelete`, don't trigger a zero downtime upgrade.
```

> Currently, only the fields managed by Ray Autoscaler, such as `replicas` and `scaleStrategy.workersToDelete`, will not trigger a zero downtime upgrade.

Reviewer 2: I think I might have asked before, but I think this could be made more precise.  

> Currently, only the fields managed by Ray Autoscaler, such as `replicas` and `scaleStrategy.workersToDelete`, will not trigger a zero downtime upgrade.

Because of "such as", this implies that there are also other fields that don't trigger it, and `replicas` and `workersToDelete` are just two examples.   If these are actually the only fields that don't trigger it, consider 

> Currently, only the fields `replicas` and `scaleStrategy.workersToDelete` will not trigger a zero downtime upgrade.  This is because these fields are managed by the Ray Autoscaler.

(I'm not sure if these are the only fields, though)

Author: Update https://github.com/ray-project/ray/pull/39074/commits/7c39cd2d2e8b9f983995c573188a0582a94ceab9

> When you update these fields, KubeRay will not propagate the update from RayService to RayCluster custom resources, so nothing will happen.

Reviewer: ```suggestion
When you update these fields, KubeRay doesn't propagate the update from RayService to RayCluster custom resources, so nothing happens.
```

```sh
# Step 8.1: Update `spec.rayClusterConfig.workerGroupSpecs[0].replicas` in the RayService YAML file from 1 to 2.

> # This field is an exception that will not trigger a zero downtime upgrade, and nothing will happen.

Reviewer: ```suggestion
# This field is an exception that doesn't trigger a zero downtime upgrade, and nothing happens.
```

kubectl apply -f ray_v1alpha1_rayservice.yaml

# Step 8.2: Check RayService CR
kubectl describe rayservices rayservice-sample
# Worker Group Specs:
#   ...
#   Replicas:  2

> # Step 8.3: Check RayCluster CR. The update is not propagated to the RayCluster CR.

Reviewer: ```suggestion
# Step 8.3: Check RayCluster CR. The update doesn't propagate to the RayCluster CR.
```

kubectl describe rayclusters $YOUR_RAY_CLUSTER
# Worker Group Specs:
#   ...
#   Replicas:  1

# Step 8.4: Update `spec.rayClusterConfig.rayVersion` to `2.100.0`.

> # This field is used to determine the Autoscaler sidecar image, and it will trigger a zero downtime upgrade.

Reviewer: ```suggestion
# This field determines the Autoscaler sidecar image, and triggers a zero downtime upgrade.
```

kubectl apply -f ray_v1alpha1_rayservice.yaml

# Step 8.5: List all RayCluster custom resources in the `default` namespace.

> # Note that the new RayCluster is created based on the updated RayService config, so it will have 2 workers.

Reviewer: ```suggestion
# Note that the new RayCluster is created based on the updated RayService config to have 2 workers.
```

kubectl get raycluster

# NAME                                 DESIRED WORKERS   AVAILABLE WORKERS   STATUS   AGE
# rayservice-sample-raycluster-6mj28   1                 1                   ready    142m
# rayservice-sample-raycluster-sjj67   2                 2                   ready    44s

> # Step 8.6: Wait for the old RayCluster to be terminated.

Reviewer: ```suggestion
# Step 8.6: Wait for the old RayCluster terminate.
```

# Step 8.7: Submit a request to the fruit stand app via the same serve service.
curl -X POST -H 'Content-Type: application/json' rayservice-sample-serve-svc:8000/fruit/ -d '["MANGO", 2]'
# [Expected output]: 8
```

> ### Another two possible scenarios that will trigger a new RayCluster preparation

Reviewer: ```suggestion
### Other possible scenarios that trigger a new RayCluster preparation
```

> Note: The following behavior is for KubeRay v0.6.2 or newer.

> For older versions, refer to [kuberay#1293](https://github.com/ray-project/kuberay/pull/1293) for more details.

Reviewer: ```suggestion
For older versions, see [kuberay#1293](https://github.com/ray-project/kuberay/pull/1293) for more details.
```

> Not only will the zero downtime upgrade trigger a new RayCluster preparation, but KubeRay will also trigger it if it considers a RayCluster unhealthy.

Reviewer: ```suggestion
KubeRay also triggers a new RayCluster preparation if it considers a RayCluster unhealthy.
```

In the RayService, KubeRay can mark a RayCluster as unhealthy in two possible scenarios.

* Case 1: The KubeRay operator can't connect to the dashboard agent on the head Pod for more than the duration defined by the `deploymentUnhealthySecondThreshold` parameter. Both the default value and values in sample YAML files of `deploymentUnhealthySecondThreshold` are 300 seconds.

> * Case 2: The KubeRay operator will mark a RayCluster as unhealthy if the status of a serve application is `DEPLOY_FAILED` or `UNHEALTHY` for a duration exceeding the `serviceUnhealthySecondThreshold` parameter. Both the default value and values in sample YAML files of `serviceUnhealthySecondThreshold` are 900 seconds.

Reviewer: ```suggestion
* Case 2: The KubeRay operator marks a RayCluster as unhealthy if the status of a serve application is `DEPLOY_FAILED` or `UNHEALTHY` for a duration exceeding the `serviceUnhealthySecondThreshold` parameter. Both the default value and values in sample YAML files of `serviceUnhealthySecondThreshold` are 900 seconds.
```

After KubeRay marks a RayCluster as unhealthy, it initiates the creation of a new RayCluster. Once the new RayCluster is ready, KubeRay redirects network traffic to it, and subsequently deletes the old RayCluster.

## Step 9: Clean up the Kubernetes cluster

```sh
# Delete the RayService.
kubectl delete -f ray_v1alpha1_rayservice.yaml

# Uninstall the KubeRay operator.
helm uninstall kuberay-operator

# Delete the curl Pod.
kubectl delete pod curl
```

## Next steps

> * Check the [RayService troubleshooting guide](kuberay-raysvc-troubleshoot) if you encounter any issues.

Reviewer: ```suggestion
* See [RayService troubleshooting guide](kuberay-raysvc-troubleshoot) if you encounter any issues.
```

> * Check [Examples](kuberay-examples) for more RayService examples.

Reviewer: ```suggestion
* See [Examples](kuberay-examples) for more RayService examples.
```
